core/controllers/reader.py

```diff
@@ -691,28 +811,103 @@ def post(self, exploration_id):
         self.render_json({})
 
 
-class StateHitEventHandler(base.BaseHandler):
+class StateHitEventHandlerNormalizedPayloadDict(TypedDict):
+    """Dict representation of StateHitEventHandler's
+    normalized_payload dictionary.
+    """
+
+    session_id: str
+    exploration_version: int
+    new_state_name: Optional[str]
+    client_time_spent_in_secs: float
+    old_params: Dict[str, Union[str, int, Dict[str, str], List[str]]]
+
+
+class StateHitEventHandler(
+    base.BaseHandler[
+        StateHitEventHandlerNormalizedPayloadDict, Dict[str, str]
+    ]
+):
     """Tracks a learner hitting a new state."""
 
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
     REQUIRE_PAYLOAD_CSRF_CHECK = False
+    URL_PATH_ARGS_SCHEMAS = {
+         'exploration_id': {
+            'schema': editor.SCHEMA_FOR_EXPLORATION_ID
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'POST': {
+            'session_id': {
+                'schema': {
+                    'type': 'basestring'
+                }
+            },
+            'exploration_version': {
+                'schema': {
+                    'type': 'int',
+                    'validators': [{
+                        'id': 'is_at_least',
+                        # Version must be greater than zero.
+                        'min_value': 1
+                    }]
+                }
+            },
+            'new_state_name': {
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'has_length_at_most',
+                        'max_value': constants.MAX_STATE_NAME_LENGTH
+                    }]
+                },
+                'default_value': None
```

Comment:
Why can this be `None`?

Reply:
Because here we are handling it gracefully ( I guess) https://github.com/oppia/oppia/blob/f8a8d680c6f95dfc3c63a003484177ac06e4ce21/core/controllers/reader.py#L718

Reply:
But we are logging an exception in the case when `new_state_name` is `None`, so it seems like it is not an expected case.

Reply:
default None removed!